feat(specs): EOF: `eof_test` generates state tests, and `execute` in live devnets

marioevz opened the pull request:
## 🗒️ Description

~~REQUIRES #1220~~ Merged

### - `eof_test` produces state tests (and blockchain tests) 

Create a state test out of an `eof_test` by:

1) in the case of a run-time container, wrapping the container under test within an init container and sending a contract-creating transaction
2) in the case of an init container, the container is wrapped in a factory container (which would execute the initcode to create a contract) and the factory container is wrapped within an init container, which is finally sent as a contract-creating transaction

### - State test generated from EOF tests have extra markers

All state tests generated automatically from the `eof_test` spec now have a different test ID, along with an extra marker applied.

The test ID will show `state_test_from_eof_test` instead of simply `state_test`, e.g.:
`test_swapn_on_max_stack[fork_Osaka-state_test_from_eof_test-swapn_operand_0]`
intead of:
`test_swapn_on_max_stack[fork_Osaka-state_test-swapn_operand_0]`

Also, all tests will have the extra markers `state_test_from_eof_test` or `blockchain_test_from_eof_test` or `blockchain_test_engine_from_eof_test` in order to be able to fill only these specific tests.

E.g. using `-m state_test` will select all tests, including `state_test_from_eof_test` tests:

```
collected 28 items / 22 deselected / 6 selected

<Package eip663_dupn_swapn_exchange>
  <Module test_swapn.py>
    <Function test_swapn_all_valid_immediates[fork_Osaka-state_test]>
    <Function test_swapn_on_max_stack[fork_Osaka-state_test_from_eof_test-swapn_operand_0]>
    <Function test_swapn_on_max_stack[fork_Osaka-state_test_from_eof_test-swapn_operand_255]>
    <Function test_swapn_stack_underflow[fork_Osaka-state_test_from_eof_test-stack_height_0]>
    <Function test_swapn_stack_underflow[fork_Osaka-state_test_from_eof_test-stack_height_1]>
    <Function test_swapn_stack_underflow[fork_Osaka-state_test_from_eof_test-stack_height_21]>
    <Function test_swapn_stack_underflow[fork_Osaka-state_test_from_eof_test-stack_height_255]>
```

But using `-m state_test_from_eof_test` will only select state tests that have been automatically generated from an EOF test:

```
collected 28 items / 23 deselected / 5 selected

<Package eip663_dupn_swapn_exchange>
  <Module test_swapn.py>
    <Function test_swapn_on_max_stack[fork_Osaka-state_test_from_eof_test-swapn_operand_0]>
    <Function test_swapn_on_max_stack[fork_Osaka-state_test_from_eof_test-swapn_operand_255]>
    <Function test_swapn_stack_underflow[fork_Osaka-state_test_from_eof_test-stack_height_0]>
    <Function test_swapn_stack_underflow[fork_Osaka-state_test_from_eof_test-stack_height_1]>
    <Function test_swapn_stack_underflow[fork_Osaka-state_test_from_eof_test-stack_height_21]>
    <Function test_swapn_stack_underflow[fork_Osaka-state_test_from_eof_test-stack_height_255]>
```

Using `-m eof_test` will still produce the exact same tests as before.

### - `eof_test` executable in live networks

This change allows the `uv run execute` command to send EOF initcode transactions to live networks by using the same methodology described in the previous section.

The marker `transaction_post_from_eof_test` will in this case send only EOF container validation tests to the live network:

```
collected 7 items / 2 deselected / 5 selected

<Package eip663_dupn_swapn_exchange>
  <Module test_swapn.py>
    <Function test_swapn_on_max_stack[fork_Osaka-transaction_post_from_eof_test-swapn_operand_0]>
    <Function test_swapn_on_max_stack[fork_Osaka-transaction_post_from_eof_test-swapn_operand_255]>
    <Function test_swapn_stack_underflow[fork_Osaka-transaction_post_from_eof_test-stack_height_0]>
    <Function test_swapn_stack_underflow[fork_Osaka-transaction_post_from_eof_test-stack_height_1]>
    <Function test_swapn_stack_underflow[fork_Osaka-transaction_post_from_eof_test-stack_height_21]>
    <Function test_swapn_stack_underflow[fork_Osaka-transaction_post_from_eof_test-stack_height_255]>
```

## 🔗 Related Issues
<!-- Reference any related issues using the GitHub issue number (e.g., Fixes #123) -->

## ✅ Checklist
- [x] All: Set appropriate labels for the changes.
- [x] All: Considered squashing commits to improve commit history.
- [x] All: Added an entry to [CHANGELOG.md](/ethereum/execution-spec-tests/blob/main/docs/CHANGELOG.md).
- [x] All: Considered updating the online docs in the [./docs/](/ethereum/execution-spec-tests/blob/main/docs/) directory.
- [x] Tests: All converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been added to [converted-ethereum-tests.txt](/ethereum/execution-spec-tests/blob/main/converted-ethereum-tests.txt).
- [x] Tests: A PR with removal of converted JSON/YML tests from [ethereum/tests](/ethereum/tests) have been opened.
- [x] Tests: Included the type and version of evm t8n tool used to locally execute test cases:  e.g., ref with commit hash or geth 1.13.1-stable-3f40e65.
- [x] Tests: Ran `mkdocs serve` locally and verified the auto-generated docs for new tests in the [Test Case Reference](https://ethereum.github.io/execution-spec-tests/main/tests/) are correctly formatted.

spencer-tb reviewed (commented):
Framework changes LGTM!

pdobacz commented on src/ethereum_test_specs/eof.py:
```diff
@@ -265,6 +299,67 @@ def verify_result(self, result: CompletedProcess, expected_result: Result, code:
+            data=initcode,
+        )
+
+        if self.expect_exception is not None or deployed_container is None:
```
the condition `deployed_container is None`: do I understand correctly, this will only trigger when `deployed_container=None` is set explicitly on the `eof_test`, and it is in `INITCODE` mode _and_ there was no `expect_exception` set, that is to say, deployment fails, but for reasons other than eof validation?

If so, maybe this could use some documentation, at least expand the one in the PR description

marioevz replied:
I added docstrings to explain the behavior of each field:
https://github.com/ethereum/execution-spec-tests/blob/fc6e7016f245b24bf17c749b2c7dcf77690c36c3/src/ethereum_test_specs/eof.py#L145-L231

These can be seen in the documentation: https://ethereum.github.io/execution-spec-tests/main/library/ethereum_test_specs/#ethereum_test_specs.EOFTest (Will get updated once this PR is merged)

The documentation itself could be improved to list `EOFTest` as a spec type that can be used with some examples, but we can leave that for a follow up PR.

pdobacz reviewed (approved):
LGTM, just one call for expanding the docs a bit

shemnon commented:
LGTM, modulo piotr's remark.

marioevz commented:
I changed the approach of the state tests generated for init-containers, and trying to infer the deployed container from the sections worked reasonably well, with just a single test that needed to be updated by setting its `deployed_container` field.

I've updated the method to be the default and will proceed to merge.

pdobacz commented on src/ethereum_test_specs/eof.py:
```diff
@@ -142,51 +143,163 @@ def run(self, *args: str, input_value: str | None = None) -> CompletedProcess:
+        container-type section will be used. An error will be raised if the deployed container
+        cannot be inferred.
+
+    If the value is set to `None`, it is assumed that the container is invalid and the test will
```
is this actually accurate? not sth like:

> If the value is set to `None`, it is assumed that the init container will fail to deploy, despite being valid EOF, and the test will expect that no contract is created.

?
